### PR TITLE
[#22737] Re-write Go SDK data plane to support timers.

### LINKED ISSUE #21164: Ensure that the payload is consumed on Elements.Data and Elements.Timers when is_last = true
BeamFnDataInboundObserver drops the bytes instead of passing it forward on the last message: https://github.com/apache/beam/blob/cd4b7f3b1af4f51bdab1a0b1a98f94b5288c09ec/sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataInboundObserver.java#L58

data_plane.py drops the timers payload but seems to do the right thing for data:
https://github.com/apache/beam/blob/cd4b7f3b1af4f51bdab1a0b1a98f94b5288c09ec/sdks/python/apache_beam/runners/worker/data_plane.py#L343

data_mgr.go drops the data payload (timers don't seem to be supported):
https://github.com/apache/beam/blob/cd4b7f3b1af4f51bdab1a0b1a98f94b5288c09ec/sdks/go/pkg/beam/core/runtime/harness/datamgr.go#L297

Imported from Jira [BEAM-13142](https://issues.apache.org/jira/browse/BEAM-13142). Original Jira may contain additional context.
Reported by: lcwik.

### PR DESCRIPTION
**Pull request by lostluck**
Support receiving and directing timers to the *exec.DataSource from the DataManager.

This doesn't fully implement timers for #22737, and only comprises the required data plane changes. This unblocks their development however, though it includes an untested TimerWriter path, and unused callback invocations in the DataSource to support this later.

The FnAPI Data Stream multiplexes all data for all instructions onto a single channel from the Runner harness. Each message may contain Data or Timers bytes for one or more instructions executing on the SDK harness, to be decoded by the execution plane's DataSource. The Data or Timers are directed to a specfic PTransform.

The data and timers may arrive before or after the associated instruction begins execution. This means that the Data and Timers must be buffered in the SDK, until they are consumed, or their associated instruction is marked as ended.  Since a single instruction may have more than one consuming PTransform it's important to track the wanted number of "islast" signals, to compare against the number received, and only signal that all byte streams for this instruction are complete when they match.

Specific things to note: 
* In elementChan we use a mutex as well as the channel to avoid concurrent changes to the "want" and "got", but use an atomic to check if elementChan is closed. This is lighterweight than using the mutex to check the condition all the time.
* The channel index in DataSource now defaults to 0 instead of -1, and we now do a "post processing check" to validate if a split has occurred. This ensures that the expected primary and residuals are respected in the same way they were before.

Unit tests were updated to match the current possible cases, and validate the above mentioned semantics. A "read" benchmark was added as well. Additional larger scale testing was also done to ensure continued correctness of the data plane's behavior.

Also Closes #21164, ensuring that the last element is always consumed.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @riteshghorse

File: sdks/go/pkg/beam/core/runtime/exec/datasource.go
Comment (riteshghorse) [inline sdks/go/pkg/beam/core/runtime/exec/datasource.go:120]: Nice, I like this approach of passing in data and timer handling functions

**Review by riteshghorse — COMMENTED** [review]
This looks great. Just few understanding questions and minor nit

**Comment by riteshghorse** [comment]
Run GoPortable PreCommit

File: sdks/go/pkg/beam/core/runtime/harness/datamgr.go
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/harness/datamgr.go]: Done!

Basically, if a setting is functionally "permanent" and decided on construction, it's better to have a separate type instead of an if-statement. Fewer ifs == faster execution. See also my long overdue performance work on the DoFn execution hotpath, which is largely blocked until timers are done, so they can be hotpathed if appropriate.

**Comment by lostluck** [comment]
Run Go Portable PreCommit

**Comment by lostluck** [comment]
Run Go PostCommit

File: sdks/go/pkg/beam/core/runtime/harness/datamgr.go
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/harness/datamgr.go:315]: There's always a "data" PTransform that isn't part of the "additionalTransforms" list.

This is something we can refactor and clean up at a later time though. Keeping the assumption makes the diff easier to understand and reduces the scope of the change a litte.

**Comment by lostluck** [comment]
Run Go PostCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Review by riteshghorse — APPROVED** [review]

**Comment by lostluck** [comment]
Thanks for the review!

Jenkins is currently down, so I'm going to wait till at most tomorrow before merging, to get those runs in.

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run Go PostCommit

**Pull request merged**
